### PR TITLE
azurerm_ip_group - support for firewall_ids, firewall_policy_ids

### LINKED ISSUE #19843: Azure Firewall: Updating multiple IP Group objects at the same time causes firewall/policy to fail
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.3.4

### AzureRM Provider Version

2.99, 3.37.0

### Affected Resource(s)/Data Source(s)

azurerm_ip_group

### Terraform Configuration Files

```hcl
# Start with this

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-network-%d"
  location = "%s"
}

resource "azurerm_ip_group" "test1" {
  name                = "acceptanceTestIpGroup1"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  cidrs = ["172.16.240.0/20"]

  tags = {
    environment = "Production"
    cost_center = "MSFT"
  }
}

resource "azurerm_ip_group" "test2" {
  name                = "acceptanceTestIpGroup2"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  cidrs = ["172.17.240.0/20"]

  tags = {
    environment = "Production"
    cost_center = "MSFT"
  }
}

resource "azurerm_firewall_policy" "test" {
  name                = "fwpol-test-policy"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_firewall_policy_rule_collection_group" "test" {
  name               = "fwpol-test"
  firewall_policy_id = azurerm_firewall_policy.test.id
  priority           = 100

  network_rule_collection {
    name     = "network-rule-collection1"
    priority = 100
    action   = "Allow"
    rule {
      name                  = "network-rule-collection1-rule1"
      protocols             = ["TCP"]
      source_ip_groups      = [azurerm_ip_group.test1.id]
      destination_ip_groups = [azurerm_ip_group.test2.id]
      destination_ports     = ["443"]
    }
  }

  network_rule_collection {
    name     = "network-rule-collection2"
    priority = 200
    action   = "Allow"
    rule {
      name                  = "network-rule-collection1-rule1"
      protocols             = ["TCP"]
      source_ip_groups      = [azurerm_ip_group.test2.id]
      destination_ip_groups = [azurerm_ip_group.test1.id]
      destination_ports     = ["443"]
    }
  }
}

resource "azurerm_virtual_network" "test" {
  name                = "testvnet"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_subnet" "test" {
  name                 = "AzureFirewallSubnet"
  resource_group_name  = azurerm_resource_group.test.name
  virtual_network_name = azurerm_virtual_network.test.name
  address_prefixes     = ["10.0.1.0/24"]
}

resource "azurerm_public_ip" "test" {
  name                = "pip-fw"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  allocation_method   = "Static"
  sku                 = "Standard"
}

resource "azurerm_firewall" "test" {
  name                = "testfirewall"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku_name            = "AZFW_VNet"
  sku_tier            = "Standard"

  firewall_policy_id  = azurerm_firewall_policy.test.id

  ip_configuration {
    name                 = "configuration"
    subnet_id            = azurerm_subnet.test.id
    public_ip_address_id = azurerm_public_ip.test.id
  }
}





# Update to that

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-network-%d"
  location = "%s"
}

resource "azurerm_ip_group" "test1" {
  name                = "acceptanceTestIpGroup1"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  cidrs = ["172.16.240.0/20", "172.18.240.0/20"]

  tags = {
    environment = "Production"
    cost_center = "MSFT"
  }
}

resource "azurerm_ip_group" "test2" {
  name                = "acceptanceTestIpGroup2"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  cidrs = ["172.17.240.0/20", "172.19.240.0/20"]

  tags = {
    environment = "Production"
    cost_center = "MSFT"
  }
}

resource "azurerm_firewall_policy" "test" {
  name                = "fwpol-test-policy"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_firewall_policy_rule_collection_group" "test" {
  name               = "fwpol-test"
  firewall_policy_id = azurerm_firewall_policy.test.id
  priority           = 100

  network_rule_collection {
    name     = "network-rule-collection1"
    priority = 100
    action   = "Allow"
    rule {
      name                  = "network-rule-collection1-rule1"
      protocols             = ["TCP"]
      source_ip_groups      = [azurerm_ip_group.test1.id]
      destination_ip_groups = [azurerm_ip_group.test2.id]
      destination_ports     = ["443"]
    }
  }

  network_rule_collection {
    name     = "network-rule-collection2"
    priority = 200
    action   = "Allow"
    rule {
      name                  = "network-rule-collection1-rule1"
      protocols             = ["TCP"]
      source_ip_groups      = [azurerm_ip_group.test2.id]
      destination_ip_groups = [azurerm_ip_group.test1.id]
      destination_ports     = ["443"]
    }
  }
}

resource "azurerm_virtual_network" "test" {
  name                = "testvnet"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_subnet" "test" {
  name                 = "AzureFirewallSubnet"
  resource_group_name  = azurerm_resource_group.test.name
  virtual_network_name = azurerm_virtual_network.test.name
  address_prefixes     = ["10.0.1.0/24"]
}

resource "azurerm_public_ip" "test" {
  name                = "pip-fw"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  allocation_method   = "Static"
  sku                 = "Standard"
}

resource "azurerm_firewall" "test" {
  name                = "testfirewall"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku_name            = "AZFW_VNet"
  sku_tier            = "Standard"

  firewall_policy_id  = azurerm_firewall_policy.test.id

  ip_configuration {
    name                 = "configuration"
    subnet_id            = azurerm_subnet.test.id
    public_ip_address_id = azurerm_public_ip.test.id
  }
}
```

### Debug Output/Panic Output

```shell
testcase.go:110: Step 3/4 error: Error running apply: exit status 1
        
        Error: waiting for the completion of Ip Group: (Name "acceptanceTestIpGroup1" / Resource Group "acctestRG-network-230103161534034082"): Code="IpGroupsUpdateFailed" Message="Put on IP Groups acceptanceTestIpGroup1 Failed with 1 faulted referenced firewalls"
        
          with azurerm_ip_group.test1,
          on terraform_plugin_test.tf line 11, in resource "azurerm_ip_group" "test1":
          11: resource "azurerm_ip_group" "test1" {
        
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: waiting for deleting "fwpol-test" (Resource Group "acctestRG-network-230103161534034082" / Policy: "fwpol-test-policy"): Code="FirewallPolicyUpdateFailed" Message="Put on Firewall Policy fwpol-test-policy Failed with 1 faulted referenced firewalls"
```

### Expected Behaviour

IP Groups that are connected with Azure Firewall or Azure Firewall Policy resources should only update one group at a time

### Actual Behaviour

Multiple updates happen at the same time causing Azure Firewall to go into faulted state

### Steps to Reproduce

Run 1st example from the above configuration example, then update the config of both `azurerm_ip_group` objects and run apply again

### Important Factoids

_No response_

### References

From https://learn.microsoft.com/en-us/azure/firewall/overview

* Configuration updates may take five minutes on average 
* An Azure Firewall configuration update can take three to five minutes on average, and parallel updates aren't supported.
* A fix is being investigated.

### PR DESCRIPTION
**Pull request by sponte**
Resolves #19843 

PS: If accepted and merged, could this be released into `v2.99`? The reason for the ask is that we use Azure CAF for landing zones and CAF is still using `v2` of the provider.

### CODE REVIEW

**Comment by sponte** [comment]
Hello, I noticed one of the checks is failing due to using `d.Get` in the delete function - what is the alternative? to query the API to get the information required i.e. in this case to get the list of linked firewalls and firewall policies?

**Comment by sponte** [comment]
Judging by the firewall example, one needs to query the API as suspected. will refactor

image

**Comment by sponte** [comment]
Refactored and updated

File: internal/services/network/ip_group_resource.go [hunk 1]
```diff
@@ -47,6 +51,22 @@ func resourceIpGroup() *pluginsdk.Resource {
...
 
+			"firewall_ids": {
+				Type:     pluginsdk.TypeSet,
+				Computed: true,
+				Elem: &pluginsdk.Schema{
+					Type: pluginsdk.TypeString,
+				},
+			},
+
+			"firewall_policy_ids": {
```
Comment (katbyte) [inline internal/services/network/ip_group_resource.go:62]: could we add these to the documentation?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @sponte - just need to add the properties to the docs and i think we're good to merge!

File: internal/services/network/ip_group_resource.go (hunk 1 above)
Comment (sponte) [reply internal/services/network/ip_group_resource.go:62]: Hey no problem - sorry was under the impression the docs are auto-generated. Let me find them

Comment (sponte) [reply internal/services/network/ip_group_resource.go:62]: Done

**Comment by sponte** [comment]
Hey @katbyte - how would we go about releasing this into v2 if at all possible?

**Comment by sponte** [comment]
Hello, gentle nudge. Is there anything else required to progress this PR?

File: internal/services/network/ip_group_resource.go [hunk 2]
```diff
@@ -47,6 +51,22 @@ func resourceIpGroup() *pluginsdk.Resource {
 
 			"resource_group_name": commonschema.ResourceGroupName(),
 
+			"firewall_ids": {
+				Type:     pluginsdk.TypeSet,
```
Comment (katbyte) [inline internal/services/network/ip_group_resource.go:55]: this should be a type list
```suggestion
				Type:     pluginsdk.TypeList,
```

File: internal/services/network/ip_group_resource.go [hunk 3]
```diff
@@ -47,6 +51,22 @@ func resourceIpGroup() *pluginsdk.Resource {
...
+				Computed: true,
+				Elem: &pluginsdk.Schema{
+					Type: pluginsdk.TypeString,
+				},
+			},
+
+			"firewall_policy_ids": {
+				Type:     pluginsdk.TypeSet,
```
Comment (katbyte) [inline internal/services/network/ip_group_resource.go:63]: And this

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @sponte - i'm sorry i missed this on my first review but exported (computed only) properties should only be lists (otherwise you need to know the hash)

As for porting it back to 2.x we try not to do so and encourage people to update to the latest version of the provider. Is there a reason you cannot upgrade?

**Comment by sponte** [comment]
> 

Hello @katbyte, I addressed the type comments.

With regards to v2 - I completely understand where you're coming from. The reason for the ask, is that we use Microsoft's Azure CAF framework which is pinned to `~> 2.99` version in https://github.com/aztfmod/terraform-azurerm-caf/blob/main/main.tf - the v3 upgrade is on the roadmap for the project but it is a substantial amount of work and is likely not going to be available for a considerable amount of time.

The fix proposed in this PR would go a long way to help with v2._x_ deployments, especially for people in the similar situation to ours.

Hope it makes sense and is not too big of an ask.

Thank you,
Stan Wozniak

**Comment by katbyte** [comment]
@sponte - you are also going to need to update the handling of the properties expand/flatter:

```
Test ended in panic.

------- Stdout: -------
=== RUN   TestAccFirewallPolicy_completePremium
=== PAUSE TestAccFirewallPolicy_completePremium
=== CONT  TestAccFirewallPolicy_completePremium

------- Stderr: -------
panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 520 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/network.resourceIpGroupCreateUpdate(0xc000323400, {0x5ddac40?, 0xc002d38800})
	/opt/teamcity-agent/work/5e6516bb4d10eb66/internal/services/network/ip_group_resource.go:91 +0x12d5
```

**Comment by sponte** [comment]
Hey @katbyte , apologies, I didn't run a test before submitting last change. I have now ran the following two tests and here are their outputs:

```
ARM_TEST_LOCATION_ALT2=uksouth ARM_TEST_LOCATION_ALT=uksouth ARM_TEST_LOCATION=uksouth TF_ACC=true go test -v github.com/hashicorp/terraform-provider-azurerm/internal/services/firewall -run TestAccFirewallPolicy_completePremium
=== RUN   TestAccFirewallPolicy_completePremium
=== PAUSE TestAccFirewallPolicy_completePremium
=== CONT  TestAccFirewallPolicy_completePremium
--- PASS: TestAccFirewallPolicy_completePremium (475.19s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/firewall      476.490s
```

and 

```
ARM_TEST_LOCATION_ALT2=uksouth ARM_TEST_LOCATION_ALT=uksouth ARM_TEST_LOCATION=uksouth TF_ACC=true go test -v github.com/hashicorp/terraform-provider-azurerm/internal/services/network -run TestAccIpGroup_updateWithAttachedPolicy  -timeout 126m
=== RUN   TestAccIpGroup_updateWithAttachedPolicy
=== PAUSE TestAccIpGroup_updateWithAttachedPolicy
=== CONT  TestAccIpGroup_updateWithAttachedPolicy
--- PASS: TestAccIpGroup_updateWithAttachedPolicy (1651.22s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/network       1652.559s
```

**Review by katbyte — APPROVED** [review]
No worried @sponte - tests pass now and LGTM 🔐

**Comment by sponte** [comment]
Excellent, thank you for merging @katbyte.

I wanted to follow up on the v2 release? Would this be possible? Do you need me to do anything to help with that?

**Pull request merged**